keps/sig-node/0028-20180906-quotas-for-ephemeral-storage.md

```diff
@@ -0,0 +1,731 @@
+  place, there is never any possibility of thin provisioning without
+  use of `discard`.
+
+  If the backing file is created densely, it additionally may require
+  significant time to create if the ephemeral limit is large.
+
+* If the filesystem is mounted `nosync`, and is sparse, it is possible
+  for writes to succeed and then fail later with I/O errors when
+  synced to the backing storage.  This will lead to data corruption
+  that cannot be detected at the time of write.
+
+  This can easily be reproduced by e. g. creating a 64MB filesystem
+  and within it creating a 128MB sparse file and building a filesystem
+  on it.  When that filesystem is in turn mounted, writes to it will
+  succeed, but I/O errors will be seen in the log and the file will be
+  incomplete:
+
+```
+# mkdir /var/tmp/d1 /var/tmp/d2
+# dd if=/dev/zero of=/var/tmp/fs1 bs=4096 count=1 seek=16383
+# mkfs.ext4 /var/tmp/fs1
+# mount -o nosync -t ext4 /var/tmp/fs1 /var/tmp/d1
+# dd if=/dev/zero of=/var/tmp/d1/fs2 bs=4096 count=1 seek=32767
+# mkfs.ext4 /var/tmp/d1/fs2
+# mount -o nosync -t ext4 /var/tmp/d1/fs2 /var/tmp/d2
+# dd if=/dev/zero of=/var/tmp/d2/test bs=4096 count=24576
+  ...will normally succeed...
+# sync
+  ...fails with I/O error!...
+```
+
+* If the filesystem is mounted `sync`, all writes to it are
+  immediately committed to the backing store, and the _dd_ operation
+  above fails as soon as it fills up `/var/tmp/d1`.  However,
+  performance is drastically slowed, particularly with small writes;
+  with 1K writes, I observed performance degradation in some cases
+  exceeding three orders of magnitude.
+
+  I performed a test comparing writing 64 MB to a base (partitioned)
+  filesystem, to a loop filesystem without `sync`, and a loop
+  filesystem with _sync.  Total I/O was sufficient to run for at least
+  5 seconds in each case.  All filesystems involved were XFS.  Loop
+  filesystems were 128 MB and dense.  Times are in seconds.  The
+  erratic behavior (e. g. the 65536 case) was involved was observed
+  repeatedly, although the exact amount of time and which I/O sizes
+  were affected varied.  The underlying device was an HP EX920 1TB
+  NVMe SSD.
+
+| I/O Size | Partition | Loop w/sync | Loop w/o sync |
+| ---:     | ---:      | ---:        | ---:          |
+| 1024 | 0.104 | 0.120 | 140.390 |
+| 4096 | 0.045 | 0.077 | 21.850 |
+| 16384 | 0.045 | 0.067 | 5.550 |
+| 65536 | 0.044 | 0.061 | 20.440 |
+| 262144 | 0.043 | 0.087 | 0.545 |
+| 1048576 | 0.043 | 0.055 | 7.490 |
+| 4194304 | 0.043 | 0.053 | 0.587 |
+
+  The only potentially viable combination in my view would be a dense
+  loop filesystem without sync, but that would render any thin
+  provisioning impossible.
+
+## Infrastructure Needed [optional]
+
+* Decision: who is responsible for quota management of all volume
+  types (and especially ephemeral volumes of all types).  At present,
+  emptydir volumes are managed by the kubelet and logdirs and writable
```

Comment:
We can start a discussion with runtime vendors to set quotas for the writable layers. Prior to that, I think we can explore more on how this is expected to implement in kubelet and what to pass to the runtime. IIUC, the ephemeral storage limit  is shared among writable layers, logs, emptyDirs, etc. If that's the case, what limit are we going to pass to the runtime to set? Passing the project id may not work because the writable layer may be in a separate filesystem (do we still need to include it?). 

Maybe we can have a separate proposal for enforcement, given that this proposal already reaches 700 lines.

Reply:
We don't pass in the project ID; the quota code picks one.  We only pass in the requested quota (or -1 for monitoring only).

I will clarify this.

Reply:
BTW, I should note that as the writable layer may be in a different filesystem from the emptydir volume(s), it may not be possible to apply an enforcing quota that applies to the entire ephemeral storage.

For example, if there is an ephemeral limit of 10 MB on the pod, and the writable layer is on a different filesystem from the emptydir volume(s), we cannot enforce a quota of 10 MB on the total.  We can enforce a limit of 10 MB on the emptydir volume(s) and a separate limit of 10 MB on the writable layer in this case.  That's inherent in the fact that quotas are always per-filesystem.

Reply:
There is already a section "Selecting a Project ID" that desribes how project IDs are selected.

Reply:
The (non)interaction with eviction is listed under non-goals.